File: 410-split-array-largest-sum/split-array-largest-sum.cpp
class Solution {
public:
    //Optimal by Binary Search TC-O(n* log(max-sum of nums)) (Same Problem as Book Allocation)
    int func(vector<int> &arr,int mid){
        int num=1;
        long long sum=0;
        for(int i=0;i<arr.size();i++){
            if(arr[i]+sum<=mid){
                sum+=arr[i];
            }else{
                num+=1;
                sum=arr[i];
            }
        }
        return num;
    }
    int splitArray(vector<int>& nums, int k) {
        int n=nums.size();
        int low=0;
        int high=0;
        for(auto i:nums){
            low=max(low,i);
            high+=i;
        }
        while(low<=high){
            int mid=(low+high)/2;
            if(func(nums,mid)>k){
                low=mid+1;
            }else{
                high=mid-1;
            }
        }
        return low;
    }
};

File: 410-split-array-largest-sum/split-array-largest-sum.py
class Solution:
    def splitArray(self, nums: list[int], k: int) -> int:
        n = len(nums)
        lo = max(nums)
        hi = 10**9
        while lo <= hi:
            mid = (lo+hi)//2
            s = 0
            cnt = 0
            for i in range(n):
                if s+nums[i]<=mid:
                    s += nums[i]
                else:
                    s = nums[i]
                    cnt += 1
            if s > 0:
                cnt += 1
            if cnt > k:
                lo = mid + 1
            else:
                hi = mid - 1
        return lo